docs/architecture/014-userids.md

```diff
@@ -0,0 +1,198 @@
+   - If a user was not found
+     - Create a new internal user id and store `attr: value` in its attributes.
+     - Set the current session's identity to the just created internal user id
+
+## User identity resolution
+
+When somebody uses the userid `xyz` in an API call, REMS searches for the user like this:
+
+1. If there is a user with a REMS internal user id `xyz`, use that user
+2. Iterate through every attribute `attr` in `:oidc-user-attributes`, in order
+   1. If there is a user with `attr: xyz`, use that user
+   2. If not, keep iterating
+3. If no user was found, or multiple users were found, return HTTP 400 Bad Request
+
+# Examples
+
+## Referring to users by external id
+
+REMS is configured to use two userid attributes:
+
+```
+:oidc-userid-attributes ["cscId" "eppn"]
+```
+
+The handler has logged in with a cscId and a user has logged in with
+an eppn. This is the `users` database table:
+
+```
+userid                           | attributes
+---------------------------------+------------------------------------------------------
+5e8edd851d2fdfbd7415232c67367cc3 | {"cscId": "handler@csc.fi", "name": "Hannah Handler"}
+b58996c504c5638798eb6b511e6f49af | {"eppn": "user@example.com", "name": "Example User"}
+```
+
+An API call impersonating the handler adds the user as a reviewer,
+referring to the user by their eppn. Note how `x-rems-user-id` must
+always be an internal REMS user id (but see Open questions).
+
+```
+POST /api/applications/request-review
+x-rems-user-id: 5e8edd851d2fdfbd7415232c67367cc3
+
+{"application-id": 1234
+ "reviewers": ["user@example.com"]}
+```
+
+REMS finds the user and the command succeeds. When the contents of the
+application are queried, the JSON document contains internal userids
+as well as external ids:
+
+```
+GET /api/applications/1234
+{"application/id": 1234,
+ "application/events":
+ [...
+  {"event/type": "application.event/review-requested",
+   "application/id": 1234,
+   "event/actor": "5e8edd851d2fdfbd7415232c67367cc3",
+   "event/actor-attributes": {"userid": "5e8edd851d2fdfbd7415232c67367cc3",
+                              "cscId" "handler@csc.fi",
+                              "name": "Hannah Handler"}
+   "application/reviewers": [{"userid": "b58996c504c5638798eb6b511e6f49af",
+                              "eppn": "user@example.com",
+                              "name": "Example User"}],
+   ...}],
+ ...}
+```
+
+## Changing an identity
+
+Problem: A user has logged in using some identity provider and created
+some applications. However now he has started logging in via another
+identity provider and the entitlements should be associated with his
+new identity.
+
+Solution: Copy the users attributes from their new `users` row to
+their old `users` row. Remove the new `users` row. Now when the user
+logs in they get their old identity, which the entitlements are
+associated with. The external id related to those entitlements has
+also changed.
```

Comment:
The entitlements and the event log would refer to the internal user id. However in the API we would output the external user id. So this would effectively change the external user id associated with their entitlements and events. This might break some consistency requirements in systems that integrate with REMS, but would be ok from the standpoint of REMS's internal consistency.

Anyway, we probably shouldn't do this to userids that have gotten entitlements. Instead the old applications should be closed and the user should just reapply for whatever is needed using their right identity.